core/src/impl/Kokkos_Core.cpp

```diff
@@ -355,29 +355,28 @@ int Kokkos::Impl::get_gpu(const InitializationSettings& settings) {
   if (settings.has_device_id()) {
     return visible_devices[settings.get_device_id()];
   }
-  // by default use the first GPU available for execution
-  // (neither device_id nor map_device_id_by are provided)
-  if (!settings.has_map_device_id_by()) {
-    return visible_devices[0];
-  }
-  // map_device_id provided
+
   // either random or round-robin assignment based on local MPI rank
-  if (!is_valid_map_device_id_by(settings.get_map_device_id_by())) {
+  if (settings.has_map_device_id_by() &&
+      !is_valid_map_device_id_by(settings.get_map_device_id_by())) {
     std::stringstream ss;
     ss << "Warning: map_device_id_by setting '"
        << settings.get_map_device_id_by() << "' is not recognized."
        << " Raised by Kokkos::initialize(int argc, char* argv[]).\n";
     Kokkos::abort(ss.str().c_str());
   }
 
-  if (settings.get_map_device_id_by() == "random") {
+  if (settings.has_map_device_id_by() &&
+      settings.get_map_device_id_by() == "random") {
     std::default_random_engine gen(get_process_id());
     std::uniform_int_distribution<int> distribution(0,
                                                     visible_devices.size() - 1);
     return visible_devices[distribution(gen)];
   }
 
-  if (settings.get_map_device_id_by() != "mpi_rank") {
+  // either map_device_id_by is not specified or it is mpi_rank
+  if (!settings.has_map_device_id_by() ||
```

Comment:
Shouldn't this be
```suggestion
  if (settings.has_map_device_id_by() &&
```

Reply:
No, this is also the code path for "the user specified nothing".  See the comment above.

Reply:
Yes, the current code aborts if either `has_map_device_id_by` hasn't been specified or it is not "mpi_rank". The one condition we want to prevent is that `has_map_device_id` is specified and it's not "mpi_rank" AFAICT.

Executing the code in its current form I run into the `abort` of not specifying anything, also see https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/9691/pipeline.

Reply:
You're right